Add MedPython to Conda

alonyado opened the pull request:
##  Adding MedPython python Package to conda


![Pepy Total Downloads](https://img.shields.io/pepy/dt/medpython)
![PyPI - License](https://img.shields.io/pypi/l/medpython)
![GitHub contributors](https://img.shields.io/github/contributors-anon/Medial-EarlySign/MR_LIBS)
![GitHub commit activity](https://img.shields.io/github/commit-activity/t/Medial-EarlySign/MR_LIBS)

[![GitHub Repo](https://img.shields.io/badge/github-repo-blue?logo=github)](https://github.com/Medial-EarlySign/MR_LIBS)
![GitHub Repo stars](https://img.shields.io/github/stars/Medial-EarlySign/MR_WIKI)


Our platform is designed to transform complex, semi-structured Electronic Medical Records (EMR) into **machine-learning-ready** data and reproducible model pipelines. The framework is optimized for the unique challenges of sparse, time-series EMR data, delivering **low memory usage** and **high-speed processing** at scale.

It was conceived as a **TensorFlow** for machine learning on medical data.

All software is now open-sourced under the MIT license. Some of the models developed by Medial EarlySign that are currently in production are available exclusively through our partners.

The framework was battle-tested in production across multiple healthcare sites and was a key component of an **award-winning** submission to the [CMS AI Health Outcomes Challenge](https://www.cms.gov/priorities/innovation/innovation-models/artificial-intelligence-health-outcomes-challenge).

## Why Use This Platform?

*   **High-Performance Processing:** Engineered for large-scale, sparse EMR time-series data where general-purpose libraries like pandas fall short.
*   **Reusable Pipelines:** Save valuable engineering time by providing shareable, tested pipelines and methods.
*   **Built-in Safeguards:** Mitigate common pitfalls like data leakage and time-series-specific overfitting.
*   **Production-Ready:** Designed for easy deployment using Docker or minimal distroless Linux images.

## Core Components

The platform is built on three key pillars:

*   **MedRepository:** A compact, efficient data repository and API for storing and accessing EMR signals. Querying categorical signals like perscriptions and diagnosis in an easy and efficient API. 
*   **MedModel:** An end-to-end machine learning pipeline that takes data from MedRepository or JSON EMR inputs to produce predictions and explainability outputs. It supports both training and inference.
*   **Medial Tools:** A suite of utilities for training, evaluation, and workflow management, including bootstrap analysis, fairness checks, and explainability.


More information on usage:

* [ETL_Infra](https://medial-earlysign.github.io/MR_Wiki/Tutorials/01.ETL%20Tutorial) - To load data into medpython data format
    - [API Documentation](https://medial-earlysign.github.io/MR_Tools)
* [AlgoMarker](https://medial-earlysign.github.io/MR_Wiki/Infrastructure%20Library/Medial%20Tools/Python/Python%20AlgoMarker%20API%20Server.md#simple-usage-example)
    - From usage in deployment/lighter API interface. Usefull for accessing prebuild models. [API Documentation](https://medial-earlysign.github.io/MR_Tools/AlgoMarker.html)
* [med library](https://medial-earlysign.github.io/MR_Wiki/Infrastructure%20Library/Medial%20Tools/Python/Examples.md)
    - Full library from training new models or using with existing models. [Complete Example: From Data to Model](https://medial-earlysign.github.io/MR_Wiki/#complete-example-from-data-to-model)
    - [API Documentation](https://medial-earlysign.github.io/MR_Tools)

## Getting Started

*   **Build a new model:** Follow the step-by-step [Tutorials](https://medial-earlysign.github.io/MR_Wiki/Tutorials/index.html) to build a model from scratch.
*   **Use an existing model:** Browse the collection of [Models](https://medial-earlysign.github.io/MR_Wiki/Models/index.html).


## Checklist:

- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
 - [x] A tarball (url) rather than a repo (e.g. git_url) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

alonyado committed:
start

alonyado committed:
bugfix

alonyado committed:
fixed

alonyado committed:
fix conda error from PR

alonyado committed:
fix for mac/windows?
Removed some older python version to have faster build

alonyado committed:
removed libboost from runtime (already included from libboost-devel).
fix spaces

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/medpython/meta.yaml```) and found it was in an excellent condition.

alonyado committed:
fix windows

alonyado committed:
Bump version to 1.0.3 and update license details

Updated version to 1.0.3 and modified license information. Compilation adjustments for macOS and windows

alonyado committed:
Modify meta.yaml for cross-platform compatibility

Updated build requirements and conditions for compatibility across platforms.

alonyado committed:
Bump medpython version to 1.0.4 - fix compilation in windows

alonyado committed:
Reduce build matrix in windows

alonyado commented:
@conda-forge/help-c-cpp ready for review

alonyado commented:
@conda-forge/help-python-c ready for review

alonyado committed:
Update to support MacOS

alonyado commented:
@conda-forge-admin please rerender

* Added MacOS Support

alonyado committed:
Update package to remove boost dependency

alonyado committed:
fix mac-os compiler

alonyado committed:
fix cross platform

alonyado committed:
update and prepare for split

alonyado committed:
test code before publish to PyPi

alonyado committed:
fix macos - retesting

alonyado committed:
rebuild after fixing macOS, added -D_LIBCPP_DISABLE_AVAILABILITY to build in Apple

alonyado committed:
fix licence file consitency

alonyado committed:
finalize build ffom source

alonyado committed:
Reduce build matrix in windows

alonyado committed:
Update to PyPi

alonyado committed:
Merge branch 'main' into medpython

alonyado committed:
Add vcomp14 dependency for Windows platform

alonyado committed:
Update to latest version, added pip check

alonyado committed:
Move to reciept format

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/medpython/recipe.yaml```) and found some lint.

Here's what I've got...


For **recipes/medpython/recipe.yaml**:

 * ❌ The top level meta key test is unexpected
 * ❌ The recipe must have some tests.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/27205798889](https://github.com/conda-forge/conda-forge-webservices/actions/runs/27205798889). Examine the logs at this URL for more detail.</sub>

alonyado committed:
fix reciept format

alonyado committed:
fix

alonyado committed:
fix names for reciept

alonyado committed:
fix test to tests

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/medpython/recipe.yaml```) and found it was in an excellent condition.

alonyado committed:
fix url

alonyado committed:
fix conditions

alonyado committed:
fix windows to use ninja

alonyado committed:
fix windows

alonyado committed:
Added missing dependency

alonyado commented:
Ready for review!  @travishathaway  

* Moved to reciept.yaml
* Added about.repository (dev_url equivalent) 
* Added medpython-etl so pip_check can be set to true

traversaro commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,82 @@
+build:
+  number: 0
+  skip:
+    - match(python, "<3.10") 
```
Can we remove this? conda-forge already does not try to build for <3.10

alonyado committed:
removed redundant python<3.10 

Removed Python version constraint for builds.

alonyado commented:
> Hi @alonyado,
> 
> Thanks for your patience on this and getting it to a working state. I think this is just about ready, but I'm still a trainee, so I need to ping the @conda-forge/help-python-c team to give a last official review.
> 
> I think one problem you're likely to come across in that review is it looks like you may be vendoring a couple of libraries in your `External` folder. This is highly discouraged on conda-forge and I encourage you to first find these vendored dependencies on conda-forge itself.
> 
> I recently went through the same struggles while adding [osm2pgsql](https://github.com/conda-forge/staged-recipes/pull/31749). Even though it may be trickier, it does really help ensure that your software is compatible with the rest of the conda-forge ecosystem.
> 
> Happy to give you more advice if you come across any challenges.

OK, I have moved all the external code. Anyway, I agree this is a better practice.

alonyado commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,81 @@
+    - if: osx
+      then: llvm-openmp
+    - ${{ stdlib("c") }}
+  host:
```
Hi @travishathaway 

Ah, that makes perfect sense! My primary goal with the FetchContent setup was to ensure that our standard PyPI wheels and standalone source builds remain entirely self-contained, so standard users don't have to manually install C++ libraries on their systems.

However, I completely understand conda-forge's policy on unvendoring and wanting to use the existing ecosystem packages.

To support both, I'll update the CMakeLists.txt to use find_package() first. This way, when the conda-forge build runs, it will detect the host dependencies (Eigen, XGBoost, etc.) and link them properly. If they aren't found (like in our isolated PyPI GitHub Actions), it will gracefully fallback to FetchContent to build statically.

alonyado committed:
Move to test from source code

alonyado committed:
Merge branch 'medpython' of github.com:Medial-EarlySign/staged-recipes into medpython

alonyado committed:
rebuild from source

alonyado committed:
Update build number and add dmlc dependency

alonyado committed:
Increment build number from 2 to 3

alonyado committed:
Added boost. clear runtime

alonyado committed:
Increment build number and add new dependencies

Updated build number and added dependencies for libxgboost and liblightgbm.

alonyado committed:
Skip windows for now

alonyado committed:
fix pypi build

alonyado commented:
Hi @travishathaway 
I have refactored the CMake build system for `medpython` to properly integrate with the Conda-Forge ecosystem or existing pacakges in the OS, shifting from static `FetchContent` fallbacks to dynamic linking.

### Key Changes
* **Dynamic Linking (5 Packages):** The CMake configuration now successfully locates and dynamically links against 5 core Conda packages: `libxgboost`, `liblightgbm`, `libboost-headers`, `eigen`, and `nlohmann_json`. I verified locally and via CI that the `FetchContent` static fallbacks are completely bypassed for these dependencies on supported platforms. The run environment contains only the minimal required runtime libraries (`libxgboost` and `liblightgbm`), while the others are treated strictly as header-only compile-time dependencies.
* **Source Configuration & PyPI Transition:** The recipe is currently pointing to the GitHub source repository rather than PyPI. The current published version on PyPI vendors these libraries internally. Building from the GitHub source today allows Conda to cleanly un-vendor them and utilize shared system libraries. The next PyPI release will be aligned with this structure.
* **Disabling Windows:** The Windows build is technically functional, but due to MSVC path resolution issues with Conda's `Library` subdirectories, it currently falls back to vendoring XGBoost via `FetchContent`. To comply strictly with Conda-Forge policies against static linking, I have disabled Windows for now rather than shipping a vendored build. I will resolve the Windows path resolution in a follow-up PR.

Please let me know if any further adjustments are needed!

alonyado commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,95 @@
+      then: vcomp14
+    - if: osx
+      then: llvm-openmp
+    - ${{ stdlib("c") }}
```
Hi @travishathaway 

​I just checked the CI logs, and I'm actually not seeing any overdepending warnings for this build.

​These are definitely required at runtime because of the linkage strategy. When I compile the wheels for PyPI, I use `FetchContent` to vendor `XGBoost` and `LightGBM` and link them statically. However, for this Conda recipe, I specifically configured CMake to use `find_package` to dynamically link against Conda's shared libraries instead.

​Because they are dynamically linked, they have to be in the run section. I verified this locally by running `ldd` on the compiled Python shared object, which shows explicit runtime bindings to Conda's `libxgboost` and `lib_lightgbm`. In fact, in one of my earlier commits, I accidentally left `libxgboost` out of the run requirements, and the test phase immediately failed with a missing shared library error.
See this commit:
Medial-EarlySign:medpython@3028ac4a8184436c4e27589fc30d97af61c857b2

​The same applies to libgomp (OpenMP) and the standard C/C++ libraries, they are dynamically linked and required at runtime for the extension to execute, so they can't just be in the host build environment.

carterbox commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,95 @@
+      then: libgomp  
+    - if: osx
+      then: llvm-openmp
+    - ${{ stdlib("c") }}
```
Please try to follow our documentations recommendations regarding OpenMP.

https://conda-forge.org/docs/maintainer/knowledge_base/#openmp

alonyado replied:
Done

carterbox commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,95 @@
+    - numpy
+    - medpython-etl
+    - libxgboost ==2.0.3
+    - liblightgbm ==4.6.0
```
liblightgbm has run_exports, so you don't need to list it again in run.

https://github.com/conda-forge/lightgbm-feedstock/blob/main/recipe/recipe.yaml

alonyado replied:
Done

carterbox commented:
> The Windows build is currently failing with a "no space left on device" error, despite previous builds succeeding.
> 
> Could you please advise on the best approach here? I can either reduce the number of pre-built binaries (trim the build matrix) to save space, or we can look into increasing the allocated storage if that is an option.
> 
> Thanks!
> 
> @conda-forge/cpp-help

Please review our config specification for options to increase disk space on the runner.

https://conda-forge.org/docs/maintainer/conda_forge_yml/#azure

carterbox commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,95 @@
+      url: https://pypi.org/packages/source/${{ name[0] }}/${{ name }}/medpython-${{ version }}.tar.gz
+      sha256: d0a985526c482a9667f65a66e7b135e579c421feac9b2f53ff1769e1a95f8e70
+    else:
+      git: https://github.com/Medial-EarlySign/medpython.git
```
Please use a tagged release on GitHub (with hash) or an sdist from PYPI, not this switchy thing.

alonyado replied:
Thanks. Added git_rev, when I build not from PyPi

carterbox replied:
Looking for something like this: https://github.com/conda-forge/nvidia-dali-python-feedstock/blob/f6a1a03a65bd9a9bf3d387a3976a0031b784a052/recipe/recipe.yaml#L18

So that we aren't using git to checkout the repository.

carterbox replied:
I assume that you aren't using the sdist from PYPI because you have made changes since the actual 1.1.2 release in order to fix conda builds.

alonyado replied:
Your assumption is spot on. I am currently using the Git source configuration solely to iterate rapidly and ensure the build chain compile cleanly in the Conda-Forge CI before tagging an official release. 

I fully agree that checking out via Git isn't the right approach for the final production recipe. 
Once the CI turns completely green here and we confirm the architecture is solid, I will:
* Publish the final, polished code to PyPI.
* Update this PR's source block to point PyPI (with a corresponding sha256 hash), by setting `use_pypi` to true.

traversaro replied:
@alonyado can you checkout this comment? Either you use the tarball and hash here? Also, if `use_pypi` is false, perhaps all the non-used code can be removed?

alonyado replied:
> [@alonyado](https://github.com/alonyado) can you checkout this comment? Either you use the tarball and hash here? Also, if `use_pypi` is false, perhaps all the non-used code can be removed?

I have publish a new and clean version using PyPi. @traversaro 
I would like to keep this Git source block in the recipe so that we can easily test future release candidates in the Conda-Forge CI before officially pushing them to PyPI.

alonyado committed:
resolve suggestions

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/medpython/recipe.yaml```) and found some lint.

Here's what I've got...


For **recipes/medpython/recipe.yaml**:

 * ❌ This recipe is using a compiler, which now requires adding a build dependence on `${{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/27631821249](https://github.com/conda-forge/conda-forge-webservices/actions/runs/27631821249). Examine the logs at this URL for more detail.</sub>

alonyado committed:
Update

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/medpython/recipe.yaml```) and found it was in an excellent condition.

alonyado committed:
fix

alonyado committed:
fix

alonyado commented:
> > The Windows build is currently failing with a "no space left on device" error, despite previous builds succeeding.
> > Could you please advise on the best approach here? I can either reduce the number of pre-built binaries (trim the build matrix) to save space, or we can look into increasing the allocated storage if that is an option.
> > Thanks!
> > @conda-forge/cpp-help
> 
> Please review our config specification for options to increase disk space on the runner.
> 
> https://conda-forge.org/docs/maintainer/conda_forge_yml/#azure

Thank you for the review and guidance! @carterbox 

Following my recent changes, there was no need to apply those specific settings, and all CI builds are now passing cleanly across all platforms.

I did want to add a quick note regarding the Windows build: you will notice we are vendoring XGBoost using CMake's `FetchContent` rather than linking the conda-forge package.

The reason is that `libxgboost==2.0.3` on Windows is compiled with MinGW, missing the MSVC-compatible `.lib` import library and `xgboostConfig.cmake`. This breaks `find_package(xgboost)` and implicit linking (`link.exe`) for standard MSVC-based Python C++ extensions.

### To Reproduce (For context)
In a standard `windows-latest` MSVC environment:
1. Install `libxgboost==2.0.3` and `cxx-compiler` via Conda.
2. Attempt standard CMake `find_package(xgboost)`.

**Reproducible Workflow & Logs:** [MR_Scripts/blob/main/.github/workflows/test_xgboost_windows.yml](https://github.com/Medial-EarlySign/MR_Scripts/blob/main/.github/workflows/test_xgboost_windows.yml)

It looks like the conda-forge maintainers resolved this missing `.lib` issue in newer XGBoost versions (like `3.2.0`). I plan to upgrade and remove this `FetchContent` workaround in a future PR, but for now, this hybrid approach successfully gets us compiling on 2.0.3!

@conda-forge/help-python-c ready for review

carterbox reviewed (changes requested):
Have you actually tested that the conda-forge mingw-built xgboost is ABI compatible with the xgboost that you build here then discard? I don't know if importing the med module is sufficient to verify that it works.

alonyado committed:
Added tests

alonyado committed:
fix recipe format

alonyado committed:
fix the source path

alonyado committed:
Added missing requests package for test

alonyado committed:
return to swig<4.3

alonyado commented:
> Have you actually tested that the conda-forge mingw-built xgboost is ABI compatible with the xgboost that you build here then discard? I don't know if importing the med module is sufficient to verify that it works.

Thank you for the feedback! @carterbox 

I have added a concrete functional test script under the `tests/` directory (`run_all.py` and an associated `model_architecture.json`). The `recipe.yaml` is now configured to run this script during the test phase. 

Regarding, the PyPi vs git source, the git source option is only for making sure the whole version works with conda, before publishing a new version. So if everything is clear, I can publish a new PyPi package and change this recipe to pull from PyPi.

@conda-forge/help-python-c ready for review

alonyado committed:
changed to PyPi release

alonyado committed:
remove swig limit

jaimergp commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,92 @@
+      sha256: eeee5a38889fe3a30241ac115fc20614c03a227a1653e775c54a0af503f6b8f4
+    else:
+      git: https://github.com/Medial-EarlySign/medpython.git
+      rev: "afc5f4992e98c97633b2175aaa3042db7ff57a57"
```
```suggestion

package:
  name: medpython
  version: ${{ version }}

source:
  url: https://pypi.org/packages/source/m/medpython/medpython-${{ version }}.tar.gz
  sha256: eeee5a38889fe3a30241ac115fc20614c03a227a1653e775c54a0af503f6b8f4
```

Let's remove this part.

jaimergp commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,92 @@
+context:
+  name: "medpython"
```
```suggestion
```

jaimergp commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,92 @@
+      then: set "CXXFLAGS=%CXXFLAGS% /openmp:llvm"
+
+    - if: not use_pypi
+      then: cd Internal/MedPyExport/generate_binding
```
```suggestion
    - cd Internal/MedPyExport/generate_binding
```

alonyado replied:
There is no need for cd, removed.

jaimergp commented on recipes/medpython/recipe.yaml:
```diff
@@ -0,0 +1,92 @@
+    - numpy
+    - medpython-etl
+    - libxgboost ==2.0.3
+
```
```suggestion
```

alonyado committed:
Apply suggestions from code review

Co-authored-by: jaimergp <jaimergp@users.noreply.github.com>

alonyado committed:
Remove cd

alonyado commented:
Hi @carterbox , apologies for the ping! Just checking in since this PR still has a "Changes Requested" block on it. I believe I've addressed all the feedback, and the recipe is now fully cleaned up and pointing to PyPI.

​Could you take a quick look when you have a moment and let me know if we are good to clear the block? Thanks!
